crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -174,6 +174,102 @@ void HMAC_verify_service_indicator(const EVP_MD *evp_md) {
   }
 }
 
+void DigestSign_verify_service_indicator(const EVP_MD_CTX *ctx) {
+  if(ctx->pctx->pmeth->pkey_id == EVP_PKEY_RSA ||
+      ctx->pctx->pmeth->pkey_id == EVP_PKEY_RSA_PSS) {
+    // SHA1 and 1024 bit keys are not approved for RSA signature generation.
+    // SHA2-224, SHA2-256, SHA2-384, SHA2-512 with 2048, 3072, and 4096 bit keys
+    // are approved for signature generation.
+    if (EVP_PKEY_size(ctx->pctx->pkey) == 256 ||
+        EVP_PKEY_size(ctx->pctx->pkey) == 384 ||
+        EVP_PKEY_size(ctx->pctx->pkey) == 512) {
+      switch (ctx->digest->type) {
+        case NID_sha224:
+        case NID_sha256:
+        case NID_sha384:
+        case NID_sha512:
+          FIPS_service_indicator_update_state();
+          break;
+        default:
+          break;
+      }
+    }
+  } else if(ctx->pctx->pmeth->pkey_id == EVP_PKEY_EC){
+    // Curves (P-224/P-256/P-384/P-521) with SHA-1, SHA2-224, SHA2-256, SHA2-384,
+    // and SHA2-512 are approved for signature generation.
+    int curve_name = EC_GROUP_get_curve_name(ctx->pctx->pkey->pkey.ec->group);
+    switch(curve_name) {
+        case NID_secp224r1:
+        case NID_X9_62_prime256v1:
+        case NID_secp384r1:
+        case NID_secp521r1:
+            goto check_digest;
+        default:
+          break;
+    }
+    check_digest:
```

Comment:
For unsupported curves will it fall through the the switch statement and still go to check_digest? Does the default case need a return? Or would it be better to track this with an approved flag and `and` it with the result of the curve and hash check?

Reply:
I've changed the `break` to a `return` for now. But I'm interested in what an approved flag with `and` would look like?